### PR TITLE
mvcc: allow clients to assign watcher IDs

### PR DESCRIPTION
**Pull request by gyuho**
Cherry-pick https://github.com/coreos/etcd/pull/8662.
Fix https://github.com/coreos/etcd/issues/7036.

/cc @connor4312

### CODE REVIEW

**Comment by xiang90** [comment]
let us wait for the 3.3 rc release. this should go into 3.4 branch not 3.3 release branch.

**Comment by gyuho** [comment]
@xiang90 Sounds good. Will merge after 3.3 release branch cut.

**Comment by xiang90** [comment]
@gyuho let us get this merged into master?

**Comment by gyuho** [comment]
@xiang90 Yeah let me rebase first and see if test passes.
Previously, some watch tests were failing.

File: clientv3/watch.go [hunk 1]
```diff
@@ -367,7 +367,7 @@ func (w *watcher) closeStream(wgs *watchGrpcStream) {
 }
 
 func (w *watchGrpcStream) addSubstream(resp *pb.WatchResponse, ws *watcherStream) {
-	if resp.WatchId == -1 {
+	if resp.Canceled && resp.CancelReason != "" {
```
Comment (gyuho) [inline clientv3/watch.go]: @xiang90 This PR breaks backward compatibility if we want to returning the original ID from watch create request.

<= v3.3: canceled watch request response always `resp.WatchId == -1`
>= v3.4: canceled watch request response may have watch ID.

So, the way to check cancellation changes. We could keep returning -1 as watch ID on cancellation, but clients won't be able to know which one is canceled...

**Comment by gyuho** [comment]
Replaced by https://github.com/coreos/etcd/pull/9065.

**Pull request closed without merging**